packages/checkout-storefront/src/sections/CheckoutForm/useCheckoutFormValidation.ts

```diff
@@ -0,0 +1,143 @@
+  } = useAddressFormUtils(shippingAddress?.country?.code as CountryCode);
+
+  const {
+    hasAllRequiredFields: billingHasAllRequiredFields,
+    getMissingFieldsFromAddress: getMissingFieldsFromBilling,
+  } = useAddressFormUtils(billingAddress?.country?.code as CountryCode);
+
+  const getShippingMissingFieldsErrorMessage = () =>
+    getAddressMissingFieldsErrorMessage(
+      "missingFieldsInShippingAddress",
+      getMissingFieldsFromShipping(shippingAddress)
+    );
+
+  const getBillingMissingFieldsErrorMessage = () =>
+    getAddressMissingFieldsErrorMessage(
+      "missingFieldsInBillingAddress",
+      getMissingFieldsFromBilling(billingAddress)
+    );
+
+  const getAddressMissingFieldsErrorMessage = (messageKey: MessageKey, fields: AddressField[]) =>
+    `${formatMessage(messageKey)}: ${fields
+      .map((field) => formatMessage(field as MessageKey))
+      .join(", ")}`;
+
+  const ensureValidCheckout = (): boolean => {
+    let isValid = true;
+    setValue("validating", true);
+    const formData = getValues();
+    const { createAccount } = formData;
+
+    try {
+      const isLoggedIn = authenticated && checkout?.email;
+
+      if (!isLoggedIn) {
+        schema.validateSyncAt("email", formData);
+      }
+    } catch (e) {
+      const { path, type } = e as ValidationError;
+      showCustomErrors([
+        { field: path as string, code: type === "email" ? "invalid" : (type as ErrorCode) },
+      ]);
+      isValid = false;
+    }
+
+    if (createAccount) {
+      try {
+        schema.validateSyncAt("password", formData);
+      } catch ({ path, type }) {
+        showCustomErrors([{ field: path as string, code: type as ErrorCode }]);
+        isValid = false;
+      }
+    }
+
+    if (!checkout.shippingAddress) {
+      showCustomErrors([{ field: "shippingAddress", code: "required" }]);
+    }
+
+    if (!shippingHasAllRequiredFields(checkout.shippingAddress)) {
+      showCustomErrors([
+        {
+          message: getShippingMissingFieldsErrorMessage(),
+          code: "invalid",
+        },
+      ]);
+      isValid = false;
+    }
+
+    if (!checkout.billingAddress) {
+      showCustomErrors([{ field: "billingAddress", code: "required" }]);
+      isValid = false;
+    }
+
+    if (
+      !isMatchingAddress(checkout.shippingAddress, checkout.billingAddress) &&
+      !billingHasAllRequiredFields(checkout.billingAddress)
+    ) {
+      showCustomErrors([
+        {
+          message: getBillingMissingFieldsErrorMessage(),
+          code: "invalid",
+        },
+      ]);
+      isValid = false;
+    }
+
+    if (!isValidProviderSelected) {
+      showCustomErrors([{ field: "paymentProvider", code: "required" }]);
+      isValid = false;
+    }
+
+    return isValid;
+  };
+
+  const validating = watch("validating");
+
+  // needs to be a hook so react doesn't batch this
+  useEffect(() => {
+    if (validating) {
+      setValue("validating", false);
+    }
+  }, [validating]);
```

Comment:
What problem are we trying to solve here?

Reply:
@bmigirl I think we can use `flushSync` instead of `useEffect` - https://github.com/reactwg/react-18/discussions/21

Reply:
Oh yeah, right. Thanks lots, will change